tests/test_ticks.c

```diff
@@ -0,0 +1,139 @@
+// Writes the 32bit value to COP0 register 9 (count)
+#define WRITE_TICKS(val) do { asm volatile("mtc0 %0,$9"::"r"(val)); } while(0)
+
+#define TEST(name, tests) do { \
+	const int NUM_CASES = sizeof(tests) / sizeof(tests[0]); \
+	for (int i=0; i < NUM_CASES; i++) { \
+		/* move ticks a little before the target to make sure it is actually doing some waiting */ \
+		mock_tick = tests[i][1] - TICKS_FROM_MS(5); \
+		WRITE_TICKS(tests[i][0]); \
+		/* act */ \
+		name(tests[i][2]); \
+		ticks_0 = TICKS_READ(); \
+		ASSERT(state != Timeout, "Case: %u Test timed out. " #name " didn't finish on time (Ticks: %#10lX)", i, ticks_0); \
+		ASSERT(state == Test && ticks_0 >= tests[i][1], "Case: %u " #name " finished too early (Ticks: %#10lX)", i, ticks_0); \
+		ASSERT(state <= Timeout, "Case: %u Unexpected state for " #name " (Ticks: %#10lX)", i, ticks_0); \
+		/* re-sync to frame */ \
+		while (state < Timeout); \
+		/* reset state */ \
+		state = Start; \
+	} \
+} while(0)
```

Comment:
As a final read, why this must be a macro and can't simply be a function?

Reply:
😀 I have no idea. Let me change it

Reply:
Ah I remember now, I used to test the macro versions as well and probably wanted to be able to test them both without much change. It is a function now.

Reply:
Great! For me this PR is done, love it!